### PR TITLE
Scale configuration continued

### PR DESCRIPTION
**Pull request by einarhauks**
Rebased @maunsen 's excellent scale configuration branch on top of latest master. 

Improved some debug messages while calibrating and added weight and calibration button to MQTT interface

### CODE REVIEW

**Comment by einarhauks** [comment]
I'm a bit surprised to see a build failure, builds fine using Platform IO on my machine. Will try to reproduce the failure but any hints are welcomed :)

**Comment by LoQue90** [comment]
Builds also fine on my local system.
Maybe it´s the order of includes in the main sketch?
embeddedwebserver.h comes before scalevoid.h. And the the missing variables are defined in the scalevoid.h, as far es i see.
But why does it build localy..

**Comment by einarhauks** [comment]
> Builds also fine on my local system. Maybe it´s the order of includes in the main sketch? embeddedwebserver.h comes before scalevoid.h. And the the missing variables are defined in the scalevoid.h, as far es i see. But why does it build localy..

The root issue is how includes and source files are handled in this project. Having implementation in .h files and relying on the include order in main.cpp is bound to cause problems. It would be beneficial to refactor the project so that header files only contain definitions and .cpp files contain implementation. All global variables should be defined in header files, that way we wouldn't be dependent on how stuff is included in main.cpp

**Comment by einarhauks** [comment]
The branch should build now, also when userconfig has not BREWMODE not set to 2 or enabled ONLYPIDSCALE

**Comment by einarhauks** [comment]
Can you approve the workflow @LoQue90 in order to get a build?

**Comment by LoQue90** [comment]
> > Builds also fine on my local system. Maybe it´s the order of includes in the main sketch? embeddedwebserver.h comes before scalevoid.h. And the the missing variables are defined in the scalevoid.h, as far es i see. But why does it build localy..
> 
> The root issue is how includes and source files are handled in this project. Having implementation in .h files and relying on the include order in main.cpp is bound to cause problems. It would be beneficial to refactor the project so that header files only contain definitions and .cpp files contain implementation. All global variables should be defined in header files, that way we wouldn't be dependent on how stuff is included in main.cpp

thats something @kjyv @murmeltier08 @fiendie can say something about :)

File: src/MQTT.h [hunk 1]
```diff
@@ -503,21 +503,19 @@ int sendHASSIODiscoveryMsg() {
...
   DiscoveryObject currentWeight = GenerateSensorDevice("currentWeight", "Weight", "g", "weight"); 
- 
+  
   // Switch Devices 
   DiscoveryObject pidOn = GenerateSwitchDevice("pidON", "Use PID"); 
   DiscoveryObject steamON = GenerateSwitchDevice("steamON", "Steam"); 
   DiscoveryObject backflushON = GenerateSwitchDevice("backflushON", "Backflush"); 
   DiscoveryObject startUsePonM = GenerateSwitchDevice("startUsePonM", "Use PonM");
 
-  // Button Devices
```
Comment (LoQue90) [inline src/MQTT.h]: typo

File: src/MQTT.h [hunk 2]
```diff
@@ -503,21 +503,19 @@ int sendHASSIODiscoveryMsg() {
...
+  // Button Devives
```
Comment (LoQue90) [inline src/MQTT.h:513]: typo

File: src/MQTT.h [hunk 3]
```diff
@@ -503,21 +503,19 @@ int sendHASSIODiscoveryMsg() {
...
   DiscoveryObject scaleCalibrateButton = GenerateButtonDevice("calibrationON", "Calibrate Scale");
   DiscoveryObject scaleTareButton = GenerateButtonDevice("tareON", "Tare Scale");
-  
 
-// Define an array to store the DiscoveryObject instances
-  DiscoveryObject discoveryObjects[] = {
-    brewSetpoint,
+std::vector<DiscoveryObject> discoveryObjects = {
+  brewSetpoint,
```
Comment (LoQue90) [inline src/MQTT.h:518]: should follow same format until next comment

File: src/MQTT.h [hunk 4]
```diff
@@ -531,22 +529,23 @@ int sendHASSIODiscoveryMsg() {
...
-    scaleTareButton
-  };
+};
 
-  const int numDiscoveryObjects = sizeof(discoveryObjects) / sizeof(discoveryObjects[0]);
+#if (BREWMODE == 2 || ONLYPIDSCALE == 1)
+discoveryObjects.push_back(currentWeight);
+discoveryObjects.push_back(scaleCalibrateButton);
+discoveryObjects.push_back(scaleTareButton);
+#endif
```
Comment (LoQue90) [inline src/MQTT.h:543]: formating

File: src/MQTT.h [hunk 5]
```diff
@@ -503,21 +503,19 @@ int sendHASSIODiscoveryMsg() {
...
   DiscoveryObject backflushON = GenerateSwitchDevice("backflushON", "Backflush"); 
   DiscoveryObject startUsePonM = GenerateSwitchDevice("startUsePonM", "Use PonM");
 
-  // Button Devices
+  // Button Devives
   DiscoveryObject scaleCalibrateButton = GenerateButtonDevice("calibrationON", "Calibrate Scale");
   DiscoveryObject scaleTareButton = GenerateButtonDevice("tareON", "Tare Scale");
-  
 
-// Define an array to store the DiscoveryObject instances
```
Comment (LoQue90) [inline src/MQTT.h:458]: i would leave this comment, but i also see the redundance with std::vector

**Comment by LoQue90** [comment]
> Can you approve the workflow @LoQue90 in order to get a build?

i dont have rights for to approv the worklfow.
@murmeltier08 could do this

**Comment by einarhauks** [comment]
I have updated the pull request with the minor formatting comments from @LoQue90 

I would appreciate if someone could test it on their machine and / or some of the people with merge rights review it (@murmeltier08 ?) :)

**Comment by LoQue90** [comment]
i don´t have a scale in my maschine or a test setup for, sorry
as far as i know murmeltier08 is on holidays right now.

ar you allready in our new discord?
there is the user fskn who is also adding the scale to his setup, maybe you both could work together?

**Comment by murmeltier08** [comment]
i can only test without a scale :/

**Comment by FabianSperrle** [comment]
> i don´t have a scale in my maschine or a test setup for, sorry as far as i know murmeltier08 is on holidays right now.
> 
> ar you allready in our new discord? there is the user fskn who is also adding the scale to his setup, maybe you both could work together?

I tried it out yesterday, but seem to be having issues with the wiring... It entered calibration mode etc, but didn't read a weight... Busy days at the moment, but I'll come back to this and continue the exploration...

File: src/pinmapping.h [hunk 6]
```diff
@@ -25,8 +25,8 @@
 #define PIN_PRESSURESENSOR 36
 #define PIN_WATERSENSOR 23
 #define PIN_FLOWSENSOR 1
-#define PIN_HXDAT 32            // Brew scale data pin
-#define PIN_HXCLK 33            // Brew scale clock pin
+#define PIN_HXDAT 35            // Brew scale data pin
```
Comment (FabianSperrle) [inline src/pinmapping.h:28]: I think this PR should not change the pin mapping. Until I changed this back, it did not work: 

```
E (7555) gpio: GPIO can only be used as input mode
[  7569][E][esp32-hal-gpio.c:130] __pinMode(): GPIO config failed
E (7558) gpio: gpio_set_level(226): GPIO output gpio_num error
```

Even if it did not output errors, that mapping clashes with the brew switch and steam switch defined above.

File: src/EmbeddedWebserver.h [hunk 7]
```diff
@@ -267,6 +272,26 @@ void serverSetup() {
         request->redirect("/");
     });
 
+#if (ONLYPIDSCALE == 1 || BREWMODE == 2)
+    server.on("/toggleTare", HTTP_POST, [](AsyncWebServerRequest *request) {
```
Comment (FabianSperrle) [inline src/EmbeddedWebserver.h:276]: When do I have to manually tare?

**Review by FabianSperrle — CHANGES_REQUESTED** [review]
Finally had a chance to flash this on the machine again... The functionality seems to work fine (after changing the pin mapping). Calibration produces a result, and the load cells return the calibrated weight after calibration. 

I did not understand why/when I would need the tare button. 

I could not test an actual shot with the scales, because they do not work properly (two cells in one HX711 -- I guess the problems are well-known). Need to switch to a second HX711 or find a different board that works better...

File: src/EmbeddedWebserver.h (hunk 7 above)
Comment (maunsen) [reply src/EmbeddedWebserver.h:276]: Manual tare is not really needed. It is a remainder of testing and of the idea of always weighing.
The scale will auto tare in the process of brewing.

File: src/pinmapping.h (hunk 6 above)
Comment (maunsen) [reply src/pinmapping.h:28]: The pinmapping is pretty old, back when the ESP8266 was the standard, but I wanted to use an ESP32. I'm still using it solo without the clevercoffee board.
This has to be adapted to the actual Pinmapping.

**Comment by fiendie** [comment]
Commits from this PR have been added to #421

**Pull request closed without merging**